### PR TITLE
Fix the upgrade tests and use annotations from JSON

### PR DESCRIPTION
**Pull request by nandajavarma**
## Description

This PR cleans up the testing pipeline a bit more by separating phases like `integration tests` and `destroy`, to support easily different use cases. The following are the currently supported use cases:
- setup gitpod in multiple cloud providers -> run integration tests -> delete the setup
- setup gitpod in multiple cloud providers -> run integration tests -> keep the setup running as preview environment
- setup gitpod in multiple cloud providers -> keep the setup running as preview environment
- setup gitpod in multiple cloud providers -> upgrade gitpod -> run integration tests -> keep the setup running as preview environment
- setup gitpod in multiple cloud providers -> upgrade gitpod -> run integration tests -> delete the setup

The PR also includes fix from @adrienthebo to delete the resources in `gitpod` workspace that deletes the LBs created in the cloud provider and cleans up records added by `external-dns`(Thanks @adrienthebo !!)

## Related Issue(s)

Fixes #

## How to test

You can run one of the following commands:
- To run integration tests and destroy cluster(supported clusters are: `aks`, `gke`, `k3s`, `eks`):
  ```
   werft run github  -a cluster=k3s -j .werft/self-hosted-installer-tests.yaml
  ```
- To run upgrade tests and keep the setup running (supported clusters are: `aks`, `gke`, `k3s`, `eks`):
  ```
   werft run github -a preview=true -a channel=beta -a version=2022.5.2 -a upgrade=true -a cluster=k3s -j .werft/self-hosted-installer-tests.yaml
  ```
    The version above is the one from which the upgrade happens. If channel is not provided, default would be `beta`
- Create a preview environment from the unstable channel on `AKS`
  ```
   werft run github -a channel=unstable -a version=2022.5.2 -a preview=true -a cluster=aks -j .werft/self-hosted-installer-tests.yaml
  ```
## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-preview

### CODE REVIEW

File: .werft/installer-tests.ts [hunk 1]
```diff
@@ -238,60 +207,120 @@ installerTests(TEST_CONFIGURATIONS[testConfig]).catch((err) => {
...
-            // there is not point in continuing if one stage fails
-            // TODO: maybe add failable, phases
+            // there is not point in continuing if one stage fails for infra setup
+            werft.fail(`create-${cloud}-infra`, "Cluster creation failed");
             break;
         }
     }
+    werft.done(`create-${cloud}-infra`);
+
+    if (upgrade && upgrade == "true") {
```
Comment (MrSimonEmms) [inline .werft/installer-tests.ts:223]: Isn't this just the same as `if (upgrade === "true")`

```suggestion
if (upgrade === "true") {
```

File: .werft/installer-tests.ts [hunk 2]
```diff
@@ -238,60 +207,120 @@ installerTests(TEST_CONFIGURATIONS[testConfig]).catch((err) => {
...
+        // runIntegrationTests()
+
+        const upgradePhase = INFRA_PHASES["KOTS_UPGRADE"];
+        const ret = callMakeTargets(upgradePhase.phase, upgradePhase.description, upgradePhase.makeTarget);
+        if (ret) {
+            return;
+        }
+    }
+
+    if (skipTests == "true") {
```
Comment (MrSimonEmms) [inline .werft/installer-tests.ts:235]:
```suggestion
    if (skipTests === "true") {
```

File: .werft/installer-tests.ts [hunk 3]
```diff
@@ -238,60 +207,120 @@ installerTests(TEST_CONFIGURATIONS[testConfig]).catch((err) => {
...
+        console.log("Skipping integration tests");
+    } else {
+        runIntegrationTests();
+    }
+
+    // if the preview flag is set to true, the script will print the result and exits
+    if (preview && preview == "true") {
```
Comment (MrSimonEmms) [inline .werft/installer-tests.ts:242]:
```suggestion
    if (preview === "true") {
```

**Review by MrSimonEmms — APPROVED** [review]
/hold

Looks good. There's a couple of tiny nits which I think should be sorted, but I'm not going to insist on it - remove the hold when you're ready.

The only question I had was how come there's a blank file added in `install/tests/install/infra/terraform/eks/providers.tf`?

**Comment by nandajavarma** [comment]
Thanks a lot for the suggestions Simon! I removed the file that committed by mistake. I committed the code fixes you suggested. Unholding!

/unhold

**Pull request merged**